### PR TITLE
[Group] Add chip-tool group command support

### PR DESCRIPTION
**Pull request by jepenven-silabs**
#### Problem
Partial fix of #1818 

#### Change overview
Added chip-tool support to parse user group commands.

*************KNOWN ISSUE*********************
Chip-tool can only send group command to group that are already set within his own GroupDataProvider instance.
GroupId **must** be in NodeId form.
Chip-tool commands to configure itself will come in a following PR.

For now two hardcoded groups exist and can be use for testing purposes.
e.g.
For commands
`./out/debug/chip-tool onoff on 0xffffffffffff0101 1`
`./out/debug/chip-tool onoff on 0xffffffffffff0102 1`

For attribute write
`/out/debug/chip-tool onoff write on-time 0 0xffffffffffff0101 1`
`/out/debug/chip-tool onoff write on-time 0 0xffffffffffff0102 1`

#### Testing
Chip-tool running on Darwin with wireshark to sniff outgoing multicast messages.

### CODE REVIEW

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h [hunk 1]
```diff
@@ -105,6 +110,38 @@ class WriteAttribute : public ModelCommand, public chip::app::WriteClient::Callb
         return mWriteClient->SendWriteRequest(device->GetSecureSession().Value());
     }
 
+    template <class T>
+    CHIP_ERROR SendGroupCommand(chip::GroupId groupId, chip::FabricIndex fabricIndex, chip::NodeId senderNodeId,
```
Comment (mrjerryjohns) [inline examples/chip-tool/commands/clusters/WriteAttributeCommand.h:114]: This is about sending a group write, not a group command right? If so, we should be clear about the naming, since 'command' really means something different from 'write'.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h [hunk 2]
```diff
@@ -105,6 +110,38 @@ class WriteAttribute : public ModelCommand, public chip::app::WriteClient::Callb
...
+        attributePathParams.mAttributeId = attributeId;
+
+        chip::Messaging::ExchangeManager * exchangeManager = chip::app::InteractionModelEngine::GetInstance()->GetExchangeManager();
+
+        ChipLogDetail(chipTool, "Sending Write Attribute to Group %u, on Fabric %x, for cluster %u with attributeId %u", groupId,
+                      fabricIndex, clusterId, attributeId);
+
+        auto writeClient = chip::Platform::MakeUnique<chip::app::WriteClient>(exchangeManager, this, mTimedInteractionTimeoutMs);
+        VerifyOrReturnError(writeClient != nullptr, CHIP_ERROR_NO_MEMORY);
+        ReturnErrorOnFailure(writeClient->EncodeAttribute(attributePathParams, value, mDataVersion));
```
Comment (mrjerryjohns) [inline examples/chip-tool/commands/clusters/WriteAttributeCommand.h:129]: All of this logic is already present in `chip::Controller::WriteAttribute` - why are we duplicating that here?

@vivien-apple why isn't this file just using those functions, instead of duplicating all of those bits here in this file?

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 1 above)
Comment (mrjerryjohns) [inline examples/chip-tool/commands/clusters/WriteAttributeCommand.h:114]: The function above then should be forcibly error'ing out if it encounters a group session, since this new function is now responsible for group comms.

Please add comments appropriately above.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h [hunk 3]
```diff
@@ -105,6 +110,38 @@ class WriteAttribute : public ModelCommand, public chip::app::WriteClient::Callb
...
+        chip::Optional<chip::SessionHandle> session =
+            exchangeManager->GetSessionManager()->CreateGroupSession(groupId, fabricIndex, senderNodeId);
+        if (!session.HasValue())
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        ReturnErrorOnFailure(writeClient->SendWriteRequest(session.Value()));
+
+        writeClient.release();
+        exchangeManager->GetSessionManager()->RemoveGroupSession(session.Value()->AsGroupSession());
```
Comment (mrjerryjohns) [inline examples/chip-tool/commands/clusters/WriteAttributeCommand.h:140]: This is an odd ownership model - why can't the session automatically be cleaned up given that a) it's a group session and b) the number of holders of that session go to 0?

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 3 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h:140]: That cannot be done automatically given the current timeframe. An issue is open for 1.X however.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 1 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h:114]: Yes, unfortunately, I don't think that changing the name in this PR is appropriate. The renaming is confusing, yes but it should be changed in another PR as a refactor (a lot stuff will need this refactoring). Moreover this is an override of the mother class `ModelCommands` which doesn't have knowledge of all the possible command types. I'm pretty sure that renaming this will cause a lot of issues and should be done outside of this PR.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 2 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h:129]: Legacy stuff most probably. The goal was to add the feature to chip-tool so that interested party could start sending group commands/attribute write with it asap not to revisit how chip-tool works...

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 1 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h:114]: Pretty sure that if we pass a yaml test to chip-tool, we're still using the function above to send groups commands. @vivien-apple ?

File: examples/chip-tool/commands/clusters/ClusterCommand.h [hunk 4]
```diff
@@ -107,6 +112,36 @@ class ClusterCommand : public ModelCommand, public chip::app::CommandSender::Cal
...
+    CHIP_ERROR SendGroupCommand(chip::GroupId groupId, chip::FabricIndex fabricIndex, chip::NodeId senderNodeId,
+                                chip::ClusterId clusterId, chip::CommandId commandId, const T & value)
+    {
+        chip::app::CommandPathParams commandPath = { 0 /* endpoint */, groupId, clusterId, commandId,
+                                                     (chip::app::CommandPathFlags::kGroupIdValid) };
+
+        chip::Messaging::ExchangeManager * exchangeManager = chip::app::InteractionModelEngine::GetInstance()->GetExchangeManager();
+
+        ChipLogDetail(chipTool, "Sending command to Group %u, on Fabric %x, for cluster %u with commandId %u", groupId, fabricIndex,
+                      clusterId, commandId);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ClusterCommand.h:125]: Please use ChipLogFormatMEI/ChipLogValueMEI for the cluster and command ids.

Please either use %u for the fabric or prefix with 0x to make it clear it's hex.

File: examples/chip-tool/commands/clusters/ClusterCommand.h [hunk 5]
```diff
@@ -107,6 +112,36 @@ class ClusterCommand : public ModelCommand, public chip::app::CommandSender::Cal
...
+        VerifyOrReturnError(commandSender != nullptr, CHIP_ERROR_NO_MEMORY);
+        ReturnErrorOnFailure(commandSender->AddRequestDataNoTimedCheck(commandPath, value, chip::NullOptional));
+
+        chip::Optional<chip::SessionHandle> session =
+            exchangeManager->GetSessionManager()->CreateGroupSession(groupId, fabricIndex, senderNodeId);
+        if (!session.HasValue())
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        ReturnErrorOnFailure(commandSender->SendGroupCommandRequest(session.Value()));
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ClusterCommand.h:137]: On failure this is leaking the group session, no?

File: examples/chip-tool/commands/clusters/ModelCommand.cpp [hunk 6]
```diff
@@ -25,6 +25,16 @@ using namespace ::chip;
 
 CHIP_ERROR ModelCommand::RunCommand()
 {
+
+    if (IsGroupId(mNodeId))
+    {
+        ChipLogProgress(chipTool, "Sending command to group 0x%" PRIx16 " on Fabric Index 0x%" PRIx16, GroupIdFromNodeId(mNodeId),
+                        static_cast<FabricIndex>(mEndPointId));
+
+        return SendGroupCommand(GroupIdFromNodeId(mNodeId), static_cast<FabricIndex>(mEndPointId),
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ModelCommand.cpp:34]: Why are we forcing consumers to pass in a fabric index (which they have no good way to know)?  We should be getting the fabric index from `CurrentCommissioner()`.

File: examples/chip-tool/commands/clusters/ModelCommand.h [hunk 7]
```diff
@@ -34,8 +34,8 @@ class ModelCommand : public CHIPCommand
 
     void AddArguments()
     {
-        AddArgument("node-id", 0, UINT64_MAX, &mNodeId);
-        AddArgument("endpoint-id", 0, UINT16_MAX, &mEndPointId);
+        AddArgument("node-id/group-id", 0, UINT64_MAX, &mNodeId);
+        AddArgument("endpoint-id/FabricIndex", 0, UINT16_MAX, &mEndPointId);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ModelCommand.h:38]: Probably better to just do:
```suggestion
        AddArgument("endpoint-id-ignored-for-group-commands", 0, UINT16_MAX, &mEndPointId);
```
or something like that....

File: examples/chip-tool/commands/clusters/ClusterCommand.h [hunk 8]
```diff
@@ -107,6 +112,36 @@ class ClusterCommand : public ModelCommand, public chip::app::CommandSender::Cal
...
+    {
+        chip::app::CommandPathParams commandPath = { 0 /* endpoint */, groupId, clusterId, commandId,
+                                                     (chip::app::CommandPathFlags::kGroupIdValid) };
+
+        chip::Messaging::ExchangeManager * exchangeManager = chip::app::InteractionModelEngine::GetInstance()->GetExchangeManager();
+
+        ChipLogDetail(chipTool, "Sending command to Group %u, on Fabric %x, for cluster %u with commandId %u", groupId, fabricIndex,
+                      clusterId, commandId);
+
+        auto commandSender = chip::Platform::MakeUnique<chip::app::CommandSender>(this, exchangeManager, false);
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ClusterCommand.h:127]: Should be passing mTimedInteractionTimeoutMs.HasValue().

File: examples/chip-tool/commands/clusters/ClusterCommand.h [hunk 9]
```diff
@@ -107,6 +112,36 @@ class ClusterCommand : public ModelCommand, public chip::app::CommandSender::Cal
...
+        VerifyOrReturnError(commandSender != nullptr, CHIP_ERROR_NO_MEMORY);
+        ReturnErrorOnFailure(commandSender->AddRequestDataNoTimedCheck(commandPath, value, chip::NullOptional));
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ClusterCommand.h:129]: Should be passing `mTimedInteractionTimeoutMs` so we can use this to test (invalid) timed group commands.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h [hunk 10]
```diff
@@ -105,6 +110,38 @@ class WriteAttribute : public ModelCommand, public chip::app::WriteClient::Callb
...
+        VerifyOrReturnError(writeClient != nullptr, CHIP_ERROR_NO_MEMORY);
+        ReturnErrorOnFailure(writeClient->EncodeAttribute(attributePathParams, value, mDataVersion));
+
+        chip::Optional<chip::SessionHandle> session =
+            exchangeManager->GetSessionManager()->CreateGroupSession(groupId, fabricIndex, senderNodeId);
+        if (!session.HasValue())
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        ReturnErrorOnFailure(writeClient->SendWriteRequest(session.Value()));
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/WriteAttributeCommand.h:137]: Leaks the session.

File: src/transport/GroupPeerMessageCounter.cpp [hunk 11]
```diff
@@ -243,7 +243,7 @@ CHIP_ERROR GroupOutgoingCounters::Init(chip::PersistentStorageDelegate * storage
     uint32_t temp;
     CHIP_ERROR err;
     err = mStorage->SyncGetKeyValue(key.GroupControlCounter(), &temp, size);
-    if (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND)
+    if (err == CHIP_ERROR_PERSISTED_STORAGE_VALUE_NOT_FOUND || err == CHIP_ERROR_KEY_NOT_FOUND)
```
Comment (bzbarsky-apple) [inline src/transport/GroupPeerMessageCounter.cpp:246]: Please file a followup to have the API define what error code gets used, so we don't have to do this guessing game....

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: examples/chip-tool/commands/clusters/ModelCommand.h (hunk 7 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/ModelCommand.h:38]: We still need the fabricIndex for data retrieval within the GroupDataProvider since the groupId is unique within a fabric

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 4 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:125]: removed debug log

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 8 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:127]: Done

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 9 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:129]: done

File: examples/chip-tool/commands/clusters/ClusterCommand.h (hunk 5 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/ClusterCommand.h:137]: yes. fixed

File: examples/chip-tool/commands/clusters/ModelCommand.h (hunk 7 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/ModelCommand.h:38]: Done

File: examples/chip-tool/commands/clusters/ModelCommand.cpp (hunk 6 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/ModelCommand.cpp:34]: Done

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 10 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h:137]: fixed

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 1 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h:114]: If a change was made so that it can handle groupSession it most probably is because a test somewhere was using it. As the goal of the PR was to enable group command for users that will input them manually I would defer this change in a future PR.

File: examples/chip-tool/commands/clusters/ClusterCommand.h [hunk 12]
```diff
@@ -107,6 +112,34 @@ class ClusterCommand : public ModelCommand, public chip::app::CommandSender::Cal
...
+
+        chip::Optional<chip::SessionHandle> session =
+            exchangeManager->GetSessionManager()->CreateGroupSession(groupId, fabricIndex, senderNodeId);
+        if (!session.HasValue())
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        CHIP_ERROR err = commandSender->SendGroupCommandRequest(session.Value());
+
+        commandSender.release();
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/ClusterCommand.h:137]: This is going to leak `commandSender` if `SendGroupCommandRequest` failed, no?

I think the logic here needs to be more like this:
```
    CHIP_ERROR err = commandSender->SendGroupCommandRequest(session.Value());
    exchangeManager->GetSessionManager()->RemoveGroupSession(session.Value()->AsGroupSession());
    ReturnErrorOnFailure(err);

    commandSender.release();

    return CHIP_NO_ERROR;
```

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h [hunk 13]
```diff
@@ -91,7 +96,7 @@ class WriteAttribute : public ModelCommand, public chip::app::WriteClient::Callb
         ChipLogProgress(chipTool, "Sending WriteAttribute to cluster " ChipLogFormatMEI " on endpoint %" PRIu16,
                         ChipLogValueMEI(clusterId), endpointId);
         chip::app::AttributePathParams attributePathParams;
-        if (!device->GetSecureSession().Value()->IsGroupSession())
+        if (device->GetSecureSession().Value()->IsGroupSession())
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/WriteAttributeCommand.h:99]: Why this change?  This does not seem right.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h [hunk 14]
```diff
@@ -105,6 +110,38 @@ class WriteAttribute : public ModelCommand, public chip::app::WriteClient::Callb
...
+
+        chip::Optional<chip::SessionHandle> session =
+            exchangeManager->GetSessionManager()->CreateGroupSession(groupId, fabricIndex, senderNodeId);
+        if (!session.HasValue())
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        CHIP_ERROR err = writeClient->SendWriteRequest(session.Value());
+
+        writeClient.release();
```
Comment (bzbarsky-apple) [inline examples/chip-tool/commands/clusters/WriteAttributeCommand.h:139]: Again, this is leaking on failure.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 13 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h:99]: Oups.

File: examples/chip-tool/commands/clusters/WriteAttributeCommand.h (hunk 14 above)
Comment (jepenven-silabs) [reply examples/chip-tool/commands/clusters/WriteAttributeCommand.h]: done

**Review by bzbarsky-apple — APPROVED** [review]

**Review by mrjerryjohns — APPROVED** [review]

**Review by selissia — APPROVED** [review]

**Comment by jepenven-silabs** [comment]
Looks like Persistent storage is failing in the test build. Looking for a fix

**Review by woody-apple — APPROVED** [review]
Fast tracking tooling updates
